File: src/simulation/ai/llm/ResponseSchema.js
import { EVENT_TYPE, WEATHER } from "../../../config/constants.js";
import {
  POLICY_TEXT_LIMITS,
  canonicalizeAiGroupId,
  getGroupPolicyContract,
} from "../../../config/aiConfig.js";

const WEATHER_SET = new Set(Object.values(WEATHER));
const EVENT_SET = new Set(Object.values(EVENT_TYPE));

function isNumber(v) {
  return typeof v === "number" && Number.isFinite(v);
}

function isRecord(value) {
  return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

function validateWeights(value, label) {
  if (!isRecord(value)) return `${label} invalid`;
  for (const [key, weight] of Object.entries(value)) {
    if (typeof key !== "string" || key.trim().length === 0) return `${label} key invalid`;
    if (!isNumber(weight)) return `${label} weight invalid`;
  }
  return "";
}

function validateOptionalText(value, label, limit) {
  if (value === undefined) return "";
  if (typeof value !== "string") return `${label} invalid`;
  if (value.length > limit * 4) return `${label} too long`;
  return "";
}

function validateOptionalNotes(value, label = "steeringNotes") {
  if (value === undefined) return "";
  if (!Array.isArray(value)) return `${label} invalid`;
  if (value.length > POLICY_TEXT_LIMITS.maxNotes * 2) return `${label} too long`;
  for (const note of value) {
    if (typeof note !== "string") return `${label} item invalid`;
    if (note.length > POLICY_TEXT_LIMITS.note * 3) return `${label} item too long`;
  }
  return "";
}

export function validateEnvironmentDirective(input) {
  if (!input || typeof input !== "object") return { ok: false, error: "directive not object" };
  if (!WEATHER_SET.has(input.weather)) return { ok: false, error: "invalid weather" };
  if (!isNumber(input.durationSec)) return { ok: false, error: "durationSec invalid" };
  if (!isNumber(input.factionTension)) return { ok: false, error: "factionTension invalid" };
  if (!Array.isArray(input.eventSpawns)) return { ok: false, error: "eventSpawns invalid" };

  for (const spawn of input.eventSpawns) {
    if (!spawn || typeof spawn !== "object") return { ok: false, error: "event spawn invalid" };
    if (!EVENT_SET.has(spawn.type)) return { ok: false, error: "event type invalid" };
    if (!isNumber(spawn.intensity)) return { ok: false, error: "event intensity invalid" };
    if (!isNumber(spawn.durationSec)) return { ok: false, error: "event duration invalid" };
  }

  const summaryError = validateOptionalText(input.summary, "summary", POLICY_TEXT_LIMITS.summary);
  if (summaryError) return { ok: false, error: summaryError };
  const focusError = validateOptionalText(input.focus, "focus", POLICY_TEXT_LIMITS.focus);
  if (focusError) return { ok: false, error: focusError };
  const notesError = validateOptionalNotes(input.steeringNotes);
  if (notesError) return { ok: false, error: notesError };

  return { ok: true, value: input };
}

export function validateGroupPolicy(input) {
  if (!input || typeof input !== "object") return { ok: false, error: "policy payload invalid" };
  if (!Array.isArray(input.policies)) return { ok: false, error: "policies missing" };

  for (const policy of input.policies) {
    if (!policy || typeof policy !== "object") return { ok: false, error: "policy item invalid" };
    if (typeof policy.groupId !== "string" || policy.groupId.length === 0) {
      return { ok: false, error: "groupId invalid" };
    }
    if (!getGroupPolicyContract(canonicalizeAiGroupId(policy.groupId))) {
      return { ok: false, error: "groupId unsupported" };
    }

    const weightsError = validateWeights(policy.intentWeights, "intentWeights");
    if (weightsError) return { ok: false, error: weightsError };

    if (!isNumber(policy.riskTolerance)) return { ok: false, error: "riskTolerance invalid" };

    const targetsError = validateWeights(policy.targetPriorities, "targetPriorities");
    if (targetsError) return { ok: false, error: targetsError };

    if (!isNumber(policy.ttlSec)) return { ok: false, error: "ttlSec invalid" };

    const summaryError = validateOptionalText(policy.summary, "summary", POLICY_TEXT_LIMITS.summary);
    if (summaryError) return { ok: false, error: summaryError };
    const focusError = validateOptionalText(policy.focus, "focus", POLICY_TEXT_LIMITS.focus);
    if (focusError) return { ok: false, error: focusError };
    const notesError = validateOptionalNotes(policy.steeringNotes);
    if (notesError) return { ok: false, error: notesError };
  }

  if (input.stateTargets !== undefined) {
    if (!Array.isArray(input.stateTargets)) return { ok: false, error: "stateTargets invalid" };
    for (const target of input.stateTargets) {
      if (!target || typeof target !== "object") return { ok: false, error: "stateTarget item invalid" };
      if (typeof target.groupId !== "string" || target.groupId.length === 0) {
        return { ok: false, error: "stateTarget groupId invalid" };
      }
      if (!getGroupPolicyContract(canonicalizeAiGroupId(target.groupId))) {
        return { ok: false, error: "stateTarget groupId unsupported" };
      }
      if (typeof target.targetState !== "string" || target.targetState.length === 0) {
        return { ok: false, error: "stateTarget targetState invalid" };
      }
      if (!isNumber(target.priority)) return { ok: false, error: "stateTarget priority invalid" };
      if (!isNumber(target.ttlSec)) return { ok: false, error: "stateTarget ttlSec invalid" };
      if (target.reason !== undefined && typeof target.reason !== "string") {
        return { ok: false, error: "stateTarget reason invalid" };
      }
    }
  }

  return { ok: true, value: input };
}

File: project_utopia/simulation/ai/llm/response_schema.py
"""Pydantic v2 response schemas for the 4 LLM decision channels.

Replaces the hand-rolled ``ResponseSchema.js`` validator. Pydantic raises
:class:`pydantic.ValidationError` on bad input; callers MUST catch and degrade
to fallback (this is how :mod:`guardrails` provides idempotent clamping).

Wire-format: every model uses ``populate_by_name=True`` plus ``Field(alias=...)``
so JSON keys stay camelCase (``durationSec``, ``factionTension`` etc.) — the
NDJSON traces, agent-bridge bodies, and existing fixtures keep working
unchanged across the JS/Py split.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

# ---------------------------------------------------------------------------
# Enums + constants (mirror JS src/config/constants.js + aiConfig.js).
# ---------------------------------------------------------------------------

WEATHER_VALUES = ("clear", "rain", "storm", "drought", "winter")
EVENT_TYPE_VALUES = (
    "animalMigration",
    "banditRaid",
    "tradeCaravan",
    "moraleBreak",
    "diseaseOutbreak",
    "wildfire",
)
DEFENSE_POSTURE_VALUES = ("defensive", "neutral", "offensive", "aggressive")

WeatherT = Literal["clear", "rain", "storm", "drought", "winter"]
EventTypeT = Literal[
    "animalMigration",
    "banditRaid",
    "tradeCaravan",
    "moraleBreak",
    "diseaseOutbreak",
    "wildfire",
]
DefensePostureT = Literal["defensive", "neutral", "offensive", "aggressive"]


# ---------------------------------------------------------------------------
# Channel 1 — environment-director
# ---------------------------------------------------------------------------


class EventSpawn(BaseModel):
    """A single weather/event spawn directive from the environment-director."""

    model_config = ConfigDict(populate_by_name=True, extra="forbid")

    type: EventTypeT
    intensity: float = Field(ge=0.0, le=10.0)
    duration_sec: float = Field(alias="durationSec", ge=0.0, le=600.0)


class EnvironmentDirective(BaseModel):
    """The environment-director's per-tick decision.

    Constraints follow ``ResponseSchema.js::validateEnvironmentDirective``.
    Numeric ranges are enforced at the pydantic boundary so the validator
    surfaces "out-of-range" rather than letting unsafe values reach the sim.
    """

    model_config = ConfigDict(populate_by_name=True, extra="forbid")

    weather: WeatherT
    duration_sec: int = Field(alias="durationSec", ge=8, le=180)
    faction_tension: float = Field(alias="factionTension", ge=0.0, le=1.0)
    event_spawns: list[EventSpawn] = Field(
        default_factory=list,
        alias="eventSpawns",
        max_length=3,
    )
    summary: str | None = None
    focus: str | None = None
    steering_notes: list[str] = Field(
        default_factory=list,
        alias="steeringNotes",
        max_length=8,
    )


# ---------------------------------------------------------------------------
# Channel 2 — npc-policy
# ---------------------------------------------------------------------------


class GroupPolicy(BaseModel):
    """A single group's intent + target policy directive."""

    model_config = ConfigDict(populate_by_name=True, extra="allow")

    group_id: str = Field(alias="groupId", min_length=1)
    intent_weights: dict[str, float] = Field(default_factory=dict, alias="intentWeights")
    risk_tolerance: float = Field(alias="riskTolerance", ge=0.0, le=1.0)
    target_priorities: dict[str, float] = Field(
        default_factory=dict, alias="targetPriorities"
    )
    ttl_sec: float = Field(alias="ttlSec", ge=8.0, le=120.0)
    summary: str | None = None
    focus: str | None = None
    steering_notes: list[str] = Field(default_factory=list, alias="steeringNotes")

    @model_validator(mode="after")
    def _check_weights_finite(self) -> "GroupPolicy":
        for name, mapping in (
            ("intent_weights", self.intent_weights),
            ("target_priorities", self.target_priorities),
        ):
            for k, v in mapping.items():
                if not isinstance(k, str) or not k.strip():
                    raise ValueError(f"{name} key invalid: {k!r}")
                if v != v or v in (float("inf"), float("-inf")):
                    raise ValueError(f"{name}[{k}] not finite")
        return self


class StateTarget(BaseModel):
    """Optional npc-policy state-machine target directive."""

    model_config = ConfigDict(populate_by_name=True, extra="forbid")

    group_id: str = Field(alias="groupId", min_length=1)
    target_state: str = Field(alias="targetState", min_length=1)
    priority: float = Field(ge=0.0, le=1.0)
    ttl_sec: float = Field(alias="ttlSec", ge=0.0, le=120.0)
    reason: str | None = None


class NpcPolicyEnvelope(BaseModel):
    """Wrapper that the npc-policy channel returns: list of policies + targets."""

    model_config = ConfigDict(populate_by_name=True, extra="allow")

    policies: list[GroupPolicy] = Field(default_factory=list)
    state_targets: list[StateTarget] = Field(default_factory=list, alias="stateTargets")


# ---------------------------------------------------------------------------
# Channel 3 — strategic-plan
# ---------------------------------------------------------------------------


class StrategicPlan(BaseModel):
    """High-level colony plan emitted by the strategic-plan channel."""

    model_config = ConfigDict(populate_by_name=True, extra="allow")

    primary_goal: str = Field(alias="primaryGoal", min_length=1)
    constraints: list[str] = Field(default_factory=list)
    resource_budget: dict[str, float] = Field(
        default_factory=dict, alias="resourceBudget"
    )
    phase: str = "bootstrap"
    defense_posture: DefensePostureT = Field(default="neutral", alias="defensePosture")
    risk_tolerance: float | None = Field(
        default=None, alias="riskTolerance", ge=0.0, le=1.0
    )


# ---------------------------------------------------------------------------
# Channel 4 — colony-agent
# ---------------------------------------------------------------------------


class ColonyPlanStep(BaseModel):
    """One step of a colony-agent's grounded plan."""

    model_config = ConfigDict(populate_by_name=True, extra="forbid")

    action: str = Field(min_length=1)
    args: dict[str, Any] = Field(default_factory=dict)
    depends_on: list[str] = Field(default_factory=list, alias="dependsOn")


class ColonyPlan(BaseModel):
    """A colony-agent plan: an ordered list of steps + per-plan metadata."""

    model_config = ConfigDict(populate_by_name=True, extra="allow")

    steps: list[ColonyPlanStep] = Field(default_factory=list)
    rationale: str | None = None
    source: str | None = None


__all__ = [
    "ColonyPlan",
    "ColonyPlanStep",
    "DEFENSE_POSTURE_VALUES",
    "DefensePostureT",
    "EVENT_TYPE_VALUES",
    "EnvironmentDirective",
    "EventSpawn",
    "EventTypeT",
    "GroupPolicy",
    "NpcPolicyEnvelope",
    "StateTarget",
    "StrategicPlan",
    "WEATHER_VALUES",
    "WeatherT",
]
